src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -4967,6 +4968,96 @@ private static void ReportDuplicateObjectMemberInitializers(BoundExpression boun
             }
         }
 
+#nullable enable
+        private static void CheckRequiredMembersInObjectInitializer(
+            MethodSymbol constructor,
+            ImmutableArray<BoundExpression> initializers,
+            SyntaxNode creationSyntax,
+            BindingDiagnosticBag diagnostics)
+        {
+            if (!constructor.ShouldCheckRequiredMembers())
+            {
+                return;
+            }
+
+            if (constructor.ContainingType.HasRequiredMembersError)
+            {
+                // An error will be reported on the constructor if from source, or a use-site diagnostic will be reported on the use if from metadata.
+                return;
+            }
+
+            var requiredMembers = constructor.ContainingType.AllRequiredMembers.ToBuilder();
+
+            if (requiredMembers.Count == 0)
+            {
+                return;
+            }
+
+            if (initializers.IsDefaultOrEmpty)
+            {
+                reportMembers();
+                return;
+            }
+
+            foreach (var initializer in initializers)
+            {
+                var (memberSymbol, initializerExpression) = initializer is BoundAssignmentOperator assignmentOperator
+                    ? (assignmentOperator.Left switch
+                    {
+                        // Regular initializers
+                        BoundObjectInitializerMember member => member.MemberSymbol,
+                        // Attribute initializers
+                        BoundPropertyAccess propertyAccess => propertyAccess.PropertySymbol,
+                        BoundFieldAccess fieldAccess => fieldAccess.FieldSymbol,
+                        _ => null
+                    }, assignmentOperator.Right)
+                    : default;
```

Comment:
nit: expression is hard to parse. Consider just grabbing the memberSymbol here, then grabbing the right part later on when we need it.